### PR TITLE
xdiff: Use hunk_func instead of out_line

### PR DESCRIPTION
**Pull request by lewis6991**
Simpler and more efficient than parsing hunk header strings and the callback gets called much less often.

### CODE REVIEW

File: src/diff.c [hunk 1]
```diff
@@ -1643,67 +1653,81 @@ diff_read(
...
-		diffstyle = DIFF_ED;
-	    else if ((STRNCMP(line, "@@ ", 3) == 0))
-	       diffstyle = DIFF_UNIFIED;
-	    else if ((STRNCMP(line, "--- ", 4) == 0)
-		    && (vim_fgets(linebuf, LBUFLEN, fd) == 0)
-		    && (STRNCMP(line, "+++ ", 4) == 0)
-		    && (vim_fgets(linebuf, LBUFLEN, fd) == 0)
-		    && (STRNCMP(line, "@@ ", 3) == 0))
-		diffstyle = DIFF_UNIFIED;
+	    if (fd == NULL)
```
Comment (lewis6991) [inline src/diff.c:1669]: Diff from here to line 1730 is unchanged and just indented.

**Comment by brammool** [comment]
@chrisbra

**Comment by chrisbra** [comment]
very nice. So instead of calling the xdiff-callback for each line, we only call it per hunk. I like it.

Small nit:

Instead of adding a separate argument to `diff_read` like this:

```
diff_read(idx_orig, idx_new, &dio->dio_diff, dio->dio_internal);
```

I would just pass down the `dio` structure. 

Also I noticed there are several test failures, that indicates that the diffing currently does not work? So it doesn't really look ready yet? Check `Test_diff_filler()` for a simple test that fails with the expected number of filler lines.

File: src/diff.c [hunk 2]
```diff
@@ -1643,67 +1653,81 @@ diff_read(
...
+	        // {first}[,{last}]c{first}[,{last}]
+	        // {first}a{first}[,{last}]
+	        // {first}[,{last}]d{first}
+	        //
+	        // unified diff looks like this:
+	        // --- file1       2018-03-20 13:23:35.783153140 +0100
+	        // +++ file2       2018-03-20 13:23:41.183156066 +0100
+	        // @@ -1,3 +1,5 @@
+	        if (isdigit(*line))
+	    	diffstyle = DIFF_ED;
```
Comment (dpelle) [inline src/diff.c]: Indentation looks inconsistent with the indentation in else block.

**Comment by chrisbra** [comment]
so, is this ready?

**Comment by lewis6991** [comment]
Yes this should be good for another review.

I fixed all the related CI failures. The remaining look unrelated to me.

**Comment by brammool** [comment]
@chrisbra let me know if this should be included now.

**Comment by chrisbra** [comment]
Yes, can be included now. Note: I had some problems applying the patch when using the patch file: https://patch-diff.githubusercontent.com/raw/vim/vim/pull/9344.patch (git complained about some unwanted whitespace changes as well as an un-parseable diff header. Using the diff file however worked: https://patch-diff.githubusercontent.com/raw/vim/vim/pull/9344.diff without any problems.

**Pull request closed without merging**